bulk_action.go

```diff
@@ -0,0 +1,106 @@
+package opensearchtools
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// Action wraps individual actions on a BulkRequest.
+type Action interface {
+	// GetAction returns a json byte array for the action and metadata of the operation.
+	GetAction() ([]byte, error)
```

Comment:
suggestion: rename this for better clarity around what this is doing. GetXXX usually implies a getter/accessor type pattern, which is usually returning some "view" of the object. Having a Get method return an error feels a bit weird. 

Another suggestion: it looks like the only reason to to have an action interface, is to help with the serialization logic of the different action types in BulkRequest.MarshalJSON. If the Action was responsible for serializing both its action and doc with the newline, then we don't need this.



.

Reply:
I'd argue that `GetAction()` is unidiomatic all on its own.  The Go convention is to leave off the `Get` prefix and just call it `Action()`.

> If the Action was responsible for serializing both its action and doc with the newline, then we don't need this.

I think I agree with this.  Making `Action` a struct with a `MarshalJSON()` method would simplify the logic overall (assuming the JSON structure is the same for them all)

Reply:
I don't think `MarshalJSON` is applicable.
```
// Marshaler is the interface implemented by types that
// can marshal themselves into valid JSON.
```
An action is multiple lines separated by newlines. Which is not valid JSON in and of itself.